source-front/source_front/api.py

```diff
@@ -0,0 +1,203 @@
+from datetime import datetime, UTC, timedelta
+from logging import Logger
+from math import trunc
+from typing import AsyncGenerator
+from urllib.parse import urlparse, parse_qs, quote
+from estuary_cdk.capture.common import LogCursor
+from estuary_cdk.http import HTTPSession
+
+from .models import (
+    FrontResource,
+    Response,
+)
+
+
+API = "https://api2.frontapp.com"
+MAX_DATE_WINDOW_SIZE = 15
+CONVERSATIONS_LAG = 2
+
+def _dt_to_s(dt: datetime) -> float:
+    timestamp = dt.timestamp()
+    # Front timestamps can have up to 3 decimal places, so we only keep the last 3 decimals.
+    return trunc(timestamp * 1000) / 1000
+
+def _s_to_dt(s: float) -> datetime:
+    return datetime.fromtimestamp(s, tz=UTC)
+
+def _extract_page_token(url: str | None) -> str | None:
+    if url is None:
+        return None
+
+    parsed_url = urlparse(url)
+    params: dict[str, str] = parse_qs(parsed_url.query) #type: ignore
+
+    return params.get("page_token")
+
+
+async def snapshot_resources(
+        http: HTTPSession,
+        path: str,
+        log: Logger,
+) -> AsyncGenerator[FrontResource, None]:
+    url = f"{API}/{path}"
+
+    params: dict[str, str | int] = {
+        "limit": 100,
+    }
+
+    while True:
+        response = Response.model_validate_json(
+            await http.request(log, url, params=params)
+        )
+
+        resources = response.results
+
+        for resource in resources:
+            yield resource
+
+        next_page_token = _extract_page_token(response.pagination.next)
+
+        if next_page_token is None:
+            break
+
+        params.update({
+            "page_token": next_page_token,
+        })
+
+
+async def _fetch_conversations_between(
+        http: HTTPSession,
+        log: Logger,
+        start: datetime,
+        end: datetime,
+) -> AsyncGenerator[FrontResource, None]:
+    """
```

Comment:
Mega-nit here: I actually would remove this doc comment, since the function name/signature are self-explanatory. Comments have tendency to drift from their associated code as updates are made to the code and corresponding updates are not made to the comments. This isn't one where I think that's likely to happen, but as a general practice I like to keep comments only for annotating non-obvious things; a good example is to document why the code _isn't_ written in a certain way, rather than just re-stating how it is written.

Reply:
Makes sense, I'll remove it. I'm also thinking that the doc comment I put in for `fetch_conversations` falls into the same category, so I'll remove that one as well.